FLOE-206

jobara opened the pull request:
Extended the PouchDB datasource to be able to add and handle views in the database for enhanced queries. Also updated to the latest version of PouchDB.

http://issues.fluidproject.org/browse/FLOE-206

jobara committed:
FLOE-206: Upgraded to latest pouchdb

Upgraded to pouchdb v2.2.3

jobara committed:
FLOE-206: Updating tests and datasource

Updated the datasource and tests after upgrading PouchDB. Now using promises.

jobara committed:
FLOE-206: Added a createView method.

A user can now add views for querying the database against. Additionally the get method was refactored to be able to make queries to the database and the set method was refactored to move shared parts into an impl function. Tests updated and added.

jobara committed:
FLOE-206: Set can now perform POST operations

Set has been extended to be able to handle both post and put requests. If the directModel doesn't contain an id, it is assumed that the request is a POST.

cindyli commented:
Seems the new pouchDB data source breaks the metadata demo.

If you launch the metadata demo, clicking "climate change impacts" or "create new resource" triggers a javascript error: TypeError: x is null, from pouchdb.js

jobara commented:
@cindyli i wasn't able to reproduce the issue you mentioned

cindyli commented:
Yes, Justin. The error disappears after upgrading firefox.

cindyli commented on src/js/pouchDBDataSource.js:
```diff
@@ -60,33 +64,80 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
+     * and add a query property that contains an object with the necessary query parameters. If the query does
+     * not require any parameters, simply provide and empty object.
+     */
     fluid.pouchdb.dataSource.get = function (database, directModel, callback) {
```
You have a unit test for get() with "get" method. Shall we add another one to get() with "query" method so that we have an example of the case to use "query"?

jobara replied:
@cindyli there is one part of the "createView: assert view" tests, but I can add a separate one as well.

jobara committed:
FLOE-206: Added a specific unit test for query

Added a specific unit test for accessing a query via the get method.

cindyli commented:
Another issue with using the new data source with the metadata demo:
1. On the metadata demo main page, click on "climate change impacts";
2. Open firebug and go to console;
3. In the editor, type in some characters, after a few key-in, this error is reported by pouchdb.js:

(node) warning: possible EventEmitter memory leak detected. 11 listeners added. Use emitter.setMaxListeners() to increase limit.

This error can be reproduced in firefox and safari, but not in chrome.

jobara committed:
FLOE-206: Updated binding to pouchdb changes

Updated to the latest api for pouchdb changes. This seems to have taken care of the memory leak errors.w

cindyli commented:
Merged @ 9fdf2da750ca74358bb41779e7018ee3a3e75807
